### DISCUSSION TITLE
Opinions on git

### DISCUSSION THREAD

**Discussion by gor77** [red/red#5531] [General]
I was always curious about this: do people who like Red actually like using git? I've always thought of git as a tool 
that's completely out of sync with Red's mission. I mean, it's all about creating better tools. About reducing complexity. 
About  having fun.

A big part of that is creating consistent tools with a good interface. A good interface is easy to learn, and hard to forget. 
But with git, it doesn't work like that :)

A tool with a terrible interface made by kernel developer conquered the world.  
It's funny in a way because I've heard from numerous people that they don't  actually enjoy using git *at all*. 
But they keep using it.

What's your opinion? Do you prefer to use something else, like Fossil or Mercurial?

**Comment by greggirwin** [comment]
Personally, I don't love git, but there's no avoiding it today. I would be great if a simpler system was practical. I tried Fossil long ago and it was OK. Mercurial was a terrible experience for me. It's all about what the goals are. Git is overkill for all but the largest distributed projects, like MS Word is overkill for most writing needs. So we use 2% of what it can do, and then we `rebase` and `clone` when things go wrong. :^)

Where I'd *like* to see this lead is apps and tools that have version control built in, but where you don't have to deal with the details. Yes, it would have more limitations, but it could even use git or something else behind the scenes, and just hide how it works under the UI. With this people can have simple data protection and recovery built in, without having to understand version control. I think that would cover most dev needs as well, with tools in the space Red lives.

**Comment by hiiamboris** [comment]
I love git's ideas and hate the implementation :) And all the GUIs to it are even more terrible than the CLI...

I didn't enjoy learning it (even the small bits I actually need), but in the end just bound the most used commands to keyboard shortcuts and after that I can't say "I don't enjoy using git", because as with all the CLI tools the goal is to automate the invocation and put it out of sight, which I did :)

**Comment by gor77** [comment]
I thought that would be the case :) 

It seems to me that tools not adequate for the job somehow became 
the standard you are not supposed to question. It's beyond me how and why exactly did that happen. 

What baffles me is that most people blindly accept this. Is it because some people are simply thriving on complexity
or because of the (wrong) assumption that "serious" and powerful tools just have to be difficult to work with, 
I don't know. But it seems wrong, just as it seems wrong that you need to spend not hours or days but weeks and 
months to learn specific GUI toolkit. And when people see Red's way of doing GUI apps, they are like 
"nice, but... this can't be serious". 

What's wrong with the world?

**Reply by hiiamboris** [reply]
The software issue is economic in nature. Everyone is in it for money only. To cut a slice of the pie.

To design ahead, to think before doing, it requires a lot of resources, including time. By the time you finish, someone will occupy the niche already and no profits for you. Capitalism is a race where short-term thinking always wins the long-term thinking. Until the term becomes so short that the whole system collapses and reboots and starts fresh.

Indeed, economic issues can again be traced deeper, and at the root of every external crisis we find a crisis of consciousness. It makes me happy to see people who question things ;)

**Reply by greggirwin** [reply]
Of course we'd love to build a simple version control system in Red, but nobody would adopt it. So then we have to ask if doing it for ourselves, with our needs and goals, adds more value, or just factures things further and acts as a barrier to keep people out. Technically superior solutions don't win on merit, sadly, and disruptions can't be summoned on demand. Eventually, the idea of tracking all this stuff so directly will probably go away, and then only a few people care how it's actually done, at which point it's possible to create a simpler system, with less features and complexity, so others can add them back. :^)

**Reply by gor77** [reply]
@hiiamboris 
Yes, I tend to agree more and more with this. I still believe it's not always just the system at fault, often it's just human
nature. People obviously want to collaborate (FLOSS is a clear example), but many of them also want to compete. 

The problem with capitalism is that it **forces** everyone to compete. And while people who favor capitalism like to talk about "healthy competition", when it comes to making a living, I think there is nothing healthy there. Mostly it's toxic.

On the other hand, it's hard to create a paying system that actually works for those who contribute. 
The problem I see with FLOSS in general is that very often  there is a lot of unpaid work behind "free". Lots of worthy contributions are coming from individuals who do it in their free time. At the same time, we are living in a system where you are expected to cover all your basic life expenses. There are crowdfunding sites, and you can ask for donations, but there is really nothing at the system level to support this often hidden, unpaid work. 


@greggirwin 
I see, but the counter argument could be that many people are already left out. 
What I think Red needs is some kind of repository of useful pieces of code that's 
discoverable and searchable. 

For example, when I was learning loops, I was searching on how to quickly create a range of numbers
and couldn't find anything except something at [Red for Python programmers](https://github.com/red/red/wiki/A-short-introduction-to-Red-for-Python-programmers#op-1). It was useful as an example, but with no step capability. 

So I wrote a function that does what I had in mind.  
Probably it's not the most idiomatic code, and maybe there are some edge cases that I'm 
not aware of, but it works. And I was pleasantly surprised when I realized I can easily put it in foreach 
loop as it is, without thinking about iterators, protocols, methods that need to be implemented or whatever. 
And I could use two dots for a function name, so it ended up very convenient (at least to me):

```
foreach x ..[2 100 2] [
; do something with even numbers up to 100
]  
```

That's great! Friction has been reduced - at least a little bit :)

But for me, git is simply not a good platform for sharing and finding snippets like this, nor for collaboration. 
With its steep learning curve it actually adds friction to collaboration process, instead of reducing it.

Will anyone who is searching for something like that (or hopefully something more useful ;) 
find it when it's needed, if the only place it exists is my github/gitlab/sourcehut/codeberg profile? 
Myriad of platforms, and each one of them is based on a tool that doesn't care about usability.

My opinion is that versioning and collaboration should be like sending mail: you don't think about IMAP or SMTP
when you write your subject line. You don't care whether it's EHLO or HELO or whatever happens behind the scene at the protocol
level. But you do need clearly visible fields for entry, attachment and a send button. 

But git is not such tool. When the interface is badly designed, or is just an afterthought,
people are forced to dig much deeper just to understand what option exactly they need. 

To me, git is something like a drilling machine with no clearly visible operating controls, but with transparent casing so you 
basically have to look through it to see what wires you need to put together just to start drilling. And may god be with
you when you want to stop.
